### PR TITLE
Implement service resource and data source in the framework provider

### PR DESCRIPTION
**Pull request by tosuke**
Output from acceptance testing:

```
$ MACKEREL_EXPERIMENTAL_TFFRAMEWORK=1 make testacc TESTS='"^TestAcc(DataSourceMackerelService(NotMatchAnyService)?|MackerelService)$$"'
TF_ACC=1 go test -v ./mackerel/... -run "^TestAcc(DataSourceMackerelService(NotMatchAnyService)?|MackerelService)$" -timeout 120m
2024/06/12 11:54:56 [INFO] mackerel: use terraform-plugin-framework based implementation
=== RUN   TestAccDataSourceMackerelService
=== PAUSE TestAccDataSourceMackerelService
=== RUN   TestAccDataSourceMackerelServiceNotMatchAnyService
=== PAUSE TestAccDataSourceMackerelServiceNotMatchAnyService
=== RUN   TestAccMackerelService
=== PAUSE TestAccMackerelService
=== CONT  TestAccDataSourceMackerelService
=== CONT  TestAccMackerelService
=== CONT  TestAccDataSourceMackerelServiceNotMatchAnyService
--- PASS: TestAccDataSourceMackerelServiceNotMatchAnyService (2.47s)
--- PASS: TestAccDataSourceMackerelService (5.17s)
--- PASS: TestAccMackerelService (8.64s)
PASS
ok      github.com/mackerelio-labs/terraform-provider-mackerel/mackerel 9.407s
```

### CODE REVIEW

File: internal/mackerel/service_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		"too short": {
+			val:       types.StringValue("a"),
+			wantError: true,
+		},
+		"too long": {
+			val:       types.StringValue("toooooooooooooooooooo-looooooooooooooooooooooooooooooooooooooooooong"),
+			wantError: true,
+		},
+		"invalid char": {
+			val:       types.StringValue("v('ω')v"),
```
Comment (Arthur1) [inline internal/mackerel/service_test.go:33]: so cute lol

File: internal/mackerel/service_test.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+			wantError: true,
+		},
+	}
+
+	ctx := context.Background()
+	for name, tt := range cases {
+		tt := tt
```
Comment (Arthur1) [inline internal/mackerel/service_test.go:40]: [may] As long as using Go 1.22+, this reassignment may be not required.

https://tip.golang.org/doc/go1.22#:~:text=Previously%2C%20the%20variables,in%20Go%201.21.

**Review by Arthur1 — COMMENTED** [review]
The review is in process, but I have commented tentatively.

Since we merged another pull request, please fix conflict?

File: internal/validatorutil/url_test.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	t.Helper()
+
+	dp, ok := d.(diag.DiagnosticWithPath)
+	if !ok {
+		t.Errorf("expected to have a path, but got no path: %+v", d)
+		return true
+	}
+
+	if !pathExpr.Matches(dp.Path()) {
+		t.Errorf("expteted to have a path that matches to %s, but got: %+v", pathExpr.String(), dp.Path())
```
[the remark this answers is not in the document]
Comment (tosuke) [reply internal/validatorutil/url_test.go]: I fixed it: https://github.com/mackerelio-labs/terraform-provider-mackerel/pull/214/commits/543e996067cee96d779a4f9072c8835225824de4

File: internal/provider/resource_mackerel_service.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *mackerelServiceResource) Update(_ context.Context, _ resource.UpdateRequest, resp *resource.UpdateResponse) {
+	resp.Diagnostics.AddError(
+		"Unable to update Service",
+		"Mackerel services are cannot be updated in-place. Please report this issue.",
```
[the remark this answers is not in the document]
Comment (tosuke) [reply internal/provider/resource_mackerel_service.go]: fixed: https://github.com/mackerelio-labs/terraform-provider-mackerel/pull/214/commits/543e996067cee96d779a4f9072c8835225824de4

File: internal/mackerel/service.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/mackerelio/mackerel-client-go"
+)
+
+func ServiceNameValidator() validator.String {
+	return stringvalidator.All(
+		stringvalidator.LengthBetween(2, 63),
+		stringvalidator.RegexMatches(regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9-_]+$`),
```
[the remark this answers is not in the document]
Comment (tosuke) [reply internal/mackerel/service.go]: As `ServiceNameValidator` is expected to be called constant times during a provider initialization, hoisting regex compilation may not be necessary. But it is some kind of best practices and there is no reason to deviate it: https://github.com/mackerelio-labs/terraform-provider-mackerel/pull/214/commits/91ffd1d4af5ebfd3a5db377389b1b6b1d99a8645

File: internal/mackerel/service_test.go (hunk 2 above)
Comment (tosuke) [reply internal/mackerel/service_test.go]: fixed: https://github.com/mackerelio-labs/terraform-provider-mackerel/pull/214/commits/225479b53b9ca1e60ddd24f4678aae875440c259

File: internal/provider/data_source_mackerel_service.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+			"id": schema.StringAttribute{
+				Computed: true,
+			},
+			"name": schema.StringAttribute{
+				Required:    true,
+				Description: "The name of the service.",
+				Validators:  []validator.String{mackerel.ServiceNameValidator()},
+			},
+			"memo": schema.StringAttribute{
+				Computed:    true,
```
[the remark this answers is not in the document]
Comment (tosuke) [reply internal/provider/data_source_mackerel_service.go:43]: Computed attribute is "output" from the provider.
We can use `memo` without specifying the value explicitly.
```terraform
data "mackerel_service" "svc" {
  name = "service"
}
resource "some_resource" "something" {
  some_value = data.mackerel_service.svc.memo
}
```

[the remark this answers is not in the document]
Comment (tosuke) [reply internal/provider/data_source_mackerel_service.go:43]: `id` may be used "input" value in the future (then `name` and `memo` are "output"). But it is out of scope.

[the remark this answers is not in the document]
Comment (Arthur1) [reply internal/provider/data_source_mackerel_service.go:43]: I understood that the computed attribute would be unnecessary if the datasource could take a memo as input.  Thank you for the information.

**Review by Arthur1 — APPROVED** [review]
Thank you.

**Pull request merged**